src/tests/unit/common/utils/test_resolve_reference.py

```diff
@@ -0,0 +1,169 @@
+import re
+import unittest
+from unittest import mock
+
+import pytest
+
+from common.exceptions import ApplicationException, NotFoundException
+from common.utils.resolve_reference import resolve_reference
+from enums import REFERENCE_TYPES, SIMOS
+from tests.unit.mock_utils import get_mock_document_service
+
+
+class ResolveReferenceTestCase(unittest.TestCase):
+    def setUp(self):
+        self.car_rental_company = {
+            "_id": "1",
+            "type": "test_data/complex/CarRental",
+            "name": "myCarRentalCompany",
+            "cars": [
+                {
+                    "type": "test_data/complex/RentalCar",
+                    "name": "Volvo 240",
+                    "plateNumber": "123",
+                    "engine": {
+                        "address": "$2",
+                        "type": SIMOS.REFERENCE.value,
+                        "referenceType": REFERENCE_TYPES.LINK.value,
+                    },
+                }
+            ],
+            "customers": [
+                {
+                    "address": "$4",
+                    "type": SIMOS.REFERENCE.value,
+                    "referenceType": REFERENCE_TYPES.LINK.value,
+                }
+            ],
+        }
+        self.customer = {
+            "_id": "4",
+            "type": "test_data/complex/Customer",
+            "name": "Jane",
+            "car": {
+                "address": "$1.cars[0]",
+                "type": SIMOS.REFERENCE.value,
+                "referenceType": REFERENCE_TYPES.LINK.value,
+            },
+        }
+        self.engine = {
+            "_id": "2",
+            "type": "test_data/complex/EngineTest",
+            "name": "myEngine",
+            "description": "",
+            "fuelPump": {
+                "address": "$3",
+                "type": SIMOS.REFERENCE.value,
+                "referenceType": REFERENCE_TYPES.LINK.value,
+            },
+            "power": 120,
+        }
+        self.fuel_pump = {
+            "_id": "3",
+            "type": "test_data/complex/FuelPumpTest",
+            "name": "fuelPump",
+            "description": "A standard fuel pump",
+        }
+        self.document_repository = mock.Mock()
+        self.document_repository.name = "datasource"
+        self.document_repository.get = self.mock_get
+        self.document_repository.find = self.mock_find
+        self.document_service = get_mock_document_service(lambda x, y: self.document_repository)
+
+    def mock_get(self, document_id: str):
+        if document_id == "1":
+            return {**self.car_rental_company}
+        if document_id == "2":
+            return {**self.engine}
+        if document_id == "3":
+            return {**self.fuel_pump}
+        if document_id == "4":
+            return {**self.customer}
+        return None
+
+    def mock_find(self, query: dict) -> list[dict]:
+        documents: list[dict] = [self.car_rental_company, self.engine, self.fuel_pump]
```

Comment:
I think the tests would pass even if just car_rental_company was in the list, since that document is the basis for all the tests. However, I will add customer too, so it won't cause issues if someone changes the tests in the future